File: scripts/publish-inventory.mjs
#!/usr/bin/env node
/* publish-inventory.mjs — push the canonical inventory.json to the live app WITHOUT a rebuild.
   POSTs to the inventory-publish Netlify function, which stores it in Netlify Blobs; the app's
   inventory.js function then serves it to the browser on the next load.

   Env required:
     INVENTORY_PUBLISH_URL     e.g. https://<your-site>/.netlify/functions/inventory-publish
     INVENTORY_PUBLISH_SECRET  must match the same env var set in Netlify
   Usage: node scripts/publish-inventory.mjs [--in <inventory.json>] [--tenant montitrentini]
*/
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const args = process.argv.slice(2);
const getArg = (f) => { const i = args.indexOf(f); return i >= 0 ? args[i + 1] : null; };

const tenant = getArg("--tenant") || "montitrentini";
const file = getArg("--in") || path.resolve(__dirname, "../src/data/montitrentini/inventory.json");

// Credentials: env vars win; otherwise fall back to a gitignored local config so the unattended
// weekly run can authenticate without the secret living in the scheduled-task prompt.
//   scripts/.inventory-publish.json  ->  { "url": "https://<site>/.netlify/functions/inventory-publish", "secret": "..." }
let URL = process.env.INVENTORY_PUBLISH_URL;
let SECRET = process.env.INVENTORY_PUBLISH_SECRET;
if (!URL || !SECRET) {
  const cfgPath = path.resolve(__dirname, ".inventory-publish.json");
  if (fs.existsSync(cfgPath)) {
    try { const c = JSON.parse(fs.readFileSync(cfgPath, "utf8")); URL = URL || c.url; SECRET = SECRET || c.secret; } catch { /* ignore */ }
  }
}
if (!URL || !SECRET) {
  console.error("✗ No publish credentials. Set INVENTORY_PUBLISH_URL + INVENTORY_PUBLISH_SECRET env vars,");
  console.error("  or create scripts/.inventory-publish.json with { \"url\": ..., \"secret\": ... } (gitignored).");
  process.exit(1);
}
if (!fs.existsSync(file)) { console.error(`✗ inventory file not found: ${file}`); process.exit(1); }

const inventory = JSON.parse(fs.readFileSync(file, "utf8"));
const n = inventory.skus ? Object.keys(inventory.skus).length : 0;

const res = await fetch(URL, {
  method: "POST",
  headers: { "Content-Type": "application/json", "x-publish-secret": SECRET },
  body: JSON.stringify({ tenant, inventory }),
});
const body = await res.text();
if (!res.ok) { console.error(`✗ publish failed ${res.status}: ${body}`); process.exit(2); }
console.log(`✓ published ${tenant} inventory (${n} SKUs) -> live store. ${body}`);

File: scripts/publish_inventory.py
#!/usr/bin/env python3
"""publish_inventory.py — Python equivalent of scripts/publish-inventory.mjs.

POSTs the canonical inventory.json to the inventory-publish Netlify function, which
validates it and writes it to Netlify Blobs. The app's inventory.js function then
serves it to the browser on next load — NO rebuild. Exists for environments without
Node (the unattended Cowork routine). Touches ONLY inventory, never pricing.

Credentials (env wins, else gitignored scripts/.inventory-publish.json):
  INVENTORY_PUBLISH_URL     https://<site>/.netlify/functions/inventory-publish
  INVENTORY_PUBLISH_SECRET  must match the same env var set in Netlify
Usage: python3 scripts/publish_inventory.py [--in <inventory.json>] [--tenant montitrentini]
"""
import sys, os, json, urllib.request, urllib.error

SCRIPTS_DIR = os.path.dirname(os.path.abspath(__file__))
REPO = os.path.dirname(SCRIPTS_DIR)
args = sys.argv[1:]

def get_arg(flag, default=None):
    if flag in args:
        i = args.index(flag)
        return args[i + 1] if i + 1 < len(args) else default
    return default

tenant = get_arg("--tenant", "montitrentini")
file = get_arg("--in", os.path.join(REPO, "src/data/montitrentini/inventory.json"))

url = os.environ.get("INVENTORY_PUBLISH_URL")
secret = os.environ.get("INVENTORY_PUBLISH_SECRET")
if not url or not secret:
    cfg_path = os.path.join(SCRIPTS_DIR, ".inventory-publish.json")
    if os.path.exists(cfg_path):
        try:
            c = json.load(open(cfg_path, encoding="utf-8"))
            url = url or c.get("url"); secret = secret or c.get("secret")
        except Exception:
            pass
if not url or not secret:
    sys.stderr.write("✗ No publish credentials. Set INVENTORY_PUBLISH_URL + INVENTORY_PUBLISH_SECRET,\n"
                     "  or create scripts/.inventory-publish.json with { \"url\": ..., \"secret\": ... } (gitignored).\n")
    sys.exit(1)
if not os.path.exists(file):
    sys.stderr.write(f"✗ inventory file not found: {file}\n"); sys.exit(1)

inventory = json.load(open(file, encoding="utf-8"))
n = len(inventory.get("skus", {}))
payload = json.dumps({"tenant": tenant, "inventory": inventory}).encode()
req = urllib.request.Request(url, data=payload, method="POST",
    headers={"Content-Type": "application/json", "x-publish-secret": secret})
try:
    with urllib.request.urlopen(req, timeout=30) as r:
        print(f"✓ published {tenant} inventory ({n} SKUs) -> live store. HTTP {r.status}: {r.read().decode()}")
except urllib.error.HTTPError as e:
    sys.stderr.write(f"✗ publish failed {e.code}: {e.read().decode()}\n"); sys.exit(2)
except Exception as e:
    sys.stderr.write(f"✗ publish error: {e}\n"); sys.exit(2)
